server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java

```diff
@@ -544,21 +556,97 @@ public String toString() {
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        Metadata.XContentContext context = Metadata.XContentContext.valueOf(params.param(CONTEXT_MODE_PARAM, CONTEXT_MODE_API));
         builder.startObject(getId());
-        builder.field("name", getName());
-        builder.field("ephemeral_id", getEphemeralId());
-        builder.field("transport_address", getAddress().toString());
+        builder.field(KEY_NAME, getName());
+        builder.field(KEY_EPHEMERAL_ID, getEphemeralId());
+        builder.field(KEY_TRANSPORT_ADDRESS, getAddress().toString());
 
-        builder.startObject("attributes");
+        builder.startObject(KEY_ATTRIBUTES);
         for (Map.Entry<String, String> entry : attributes.entrySet()) {
             builder.field(entry.getKey(), entry.getValue());
         }
         builder.endObject();
+        if (context == Metadata.XContentContext.GATEWAY) {
+            builder.field(KEY_HOST_NAME, getHostName());
+            builder.field(KEY_HOST_ADDRESS, getHostAddress());
+            builder.field(KEY_VERSION, getVersion().toString());
+            builder.startArray(KEY_ROLES);
+            for (DiscoveryNodeRole role : roles) {
+                builder.value(role.roleName());
+            }
+            builder.endArray();
+        }
 
         builder.endObject();
         return builder;
     }
 
+    public static DiscoveryNode fromXContent(XContentParser parser, String nodeId) throws IOException {
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            parser.nextToken();
+        }
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new IllegalArgumentException("expected field name but got a " + parser.currentToken());
+        }
+        String nodeName = null;
+        String hostName = null;
+        String hostAddress = null;
+        String ephemeralId = null;
+        TransportAddress transportAddress = null;
+        Map<String, String> attributes = new HashMap<>();
+        Set<DiscoveryNodeRole> roles = new HashSet<>();
+        Version version = null;
+        String currentFieldName = parser.currentName();
+        // token should be start object at this point
+        // XContentParser.Token token = parser.nextToken();
+        // if (token != XContentParser.Token.START_OBJECT) {
+        // throw new IllegalArgumentException("expected object but got a " + token);
+        // }
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
This doesn't seem to align with your implementation of `ToXContentFragment` (which doesn't require start/end object) as well as the code on lines 595-597 which implies the `START_OBJECT` is optional.  You'd hit the end of the parser without getting the `END_OBJECT`. (Same comment on `DiscoveryNodes` class below.)

Reply:
I believe there might be some misunderstanding here. I agree that DiscoveryNode object's XContent does not need to have a start and end object, but here we are checking for the end object which is added in line 581 above

Reply:
But for `DiscoveryNodes`, this comment does make sense. I will modify the class to implement ToXContentObject.

Reply:
> I believe there might be some misunderstanding here. I agree that DiscoveryNode object's XContent does not need to have a start and end object

I'm not sure this is the desired state here. Why don't you want to extend `ToXContentObject` instead of `ToXContentFragment`?

> but here we are checking for the end object which is added in line 581 above

You are also adding a start object.  Which is fine as long as they're paired.  But you can't guarantee (based on the interface type) that you're always receiving the same object you wrote.  It's possible someone else may make a larger object and include this inline -- the purpose of `ToXContentFragment`.

You make the start object optional here:
```java
        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
            parser.nextToken();
        }
```

So if someone else trusts your object inheritance and places the object inline without start and end objects then you'd never get to the end and would keep iterating `parser.nextToken()`.  Eventually that will be null, and null won't equal end object, so inside the loop you'll throw an NPE on `token.isValue()`.

So you need to either test for a null `nextToken()` in addition to the END_OBJECT as a loop stopping condition, or you need to implement/extend `ToXContentObject` which requires start and end objects.   I suspect you probably want the latter.

Reply:
Actually, `DiscoveryNode` is part of the `DiscoveryNodes` toXContent response and getting used as a fragment like you mentioned and I plan to use `DiscoveryNodes` toXContent in ClusterState going forward. That's the reason for keeping them Fragment. Anyway, I have pushed a commit modifying the toXContent method. If you still believe we can hit a NPE in the current method. Can you please provide a sample json, I would write a unit test based on that to ensure that NPE is never thrown from our function.

Reply:
> Actually, `DiscoveryNode` is part of the `DiscoveryNodes` toXContent response and getting used as a fragment

No, it's being used as an object.  Your example in a comment below shows how it's used inline, and it includes th e `{` (START_OBJECT) and `}` (END_OBJECT).  It's a nested complete object, not a fragment.
```json
"node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
}
```
For a fragment example, see the `TransportAddress` class.  Its `toXContent` method simply adds the string value, without the object wrapper:
```java
@Override
public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
    return builder.value(toString());
}   
```
This is a trivial example, writing the `toString()` (which is what you read when parsing) but is more commonly used for creating arrays.... you'd have START_ARRAY and several objects and then END_ARRAY.  [Here is an example](https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/transport/DeleteModelResponse.java).

In your builder, you always start and end with the start object/end object, so as long as you're always parsing JSON that you created with your own `toXContent()` method, your code will work.   

So I have to be a bit creative to get malformed JSON but here's an example: The problem is that by making start object optional, you would allow malformed JSON like this to partially parse:
```json
{ [
  "node_1" : 
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
] }
```
Note the start object external to the array at the start. You read the field name and then try to read this object.  Since you've made start object optional, it jumps right in and parses name, ephemeral_id, transport_address, and attributes.   it's still looking for the `}` but doesn't find it, instead it finds the end array. With your current code it would throw an exception on an unexpected token.

If you didn't have the array and did have an enclosing outer object, you'd now be mismatched as you'd have read the closing object for something else other than your own object, and would get some other confusing parsing error that didn't really match the _real_ error, the omission of the `{` at the start of this object.

TLDR:
1. you _must_ match a required end object to a required start object.  You require one but make the other one optional.
2. you are implementing `ToXContentFragment` but you really should implement `ToXContentObject` here.

Reply:
> you are implementing ToXContentFragment but you really should implement ToXContentObject here.

https://github.com/opensearch-project/OpenSearch/blob/9b9158e4cb692dfe04315225ce2b5f9c6c453307/server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java#L69

I am not changing the implementing `ToXContentFragment` interface in the `DiscoveryNode`. It was already an implementation of the this interface and changing it to a `ToXContentObject` might causing breaking changes in our multiple API responses. As this object is already getting used as a inline object in cluster state API, and other cat APIs.

>No, it's being used as an object. Your example in a comment below shows how it's used inline, and it includes th e { (START_OBJECT) and } (END_OBJECT). It's a nested complete object, not a fragment.

Even a nested object is a fragment only as there is no start and end object around it. According to our documentation [here](https://github.com/opensearch-project/OpenSearch/blob/9b9158e4cb692dfe04315225ce2b5f9c6c453307/libs/core/src/main/java/org/opensearch/core/xcontent/ToXContentFragment.java#L36-L41), a fragment needs to have start and end object externally.
The example of you gave around fragment being used for creating array. My use case is also similar, where instead of create a field which has a `START_ARRAY`, our object has `START_OBJECT`.

The JSON you have provided is a invalid. And we don't expect to parse XContent not created from our method. Even if our object is getting used inline, like I am using in `DiscoveryNodes` class, the XContent is created from our toXContent method and object is parsed from our fromXContent method. 

> you must match a required end object to a required start object. You require one but make the other one optional

I am matching start and end object which are being added in our toXContent method. The line 590-592 you are referring, is to enable parsing of following object. This is a complete XContent object, but even if we don't receive the complete object (i.e. without start and end), we will be able to parse that inline object.
```
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```

Reply:
> I am not changing the implementing `ToXContentFragment` interface in the `DiscoveryNode`. It was already an implementation of the this interface and changing it to a `ToXContentObject` might causing breaking changes 

Agreed, you shouldn't change this.

> Even a nested object is a fragment only as there is no start and end object around it. 

Agreed.... but then you check for the end object in the line prompting this comment chain:
```java
 while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) 
```
If there is no end object, this loop will never reach its stopping condition.

> The JSON you have provided is a invalid.

I know. But the parsing error thrown would have been somewhere unrelated to the invalid JSON, in some follow-on code much more confusing to debug.  I just threw up a quick example to show that in the case you do not require a start object, and thus shouldn't require an end object (because they should match) you end up with the code as it was written hitting some later end object.  Or possibly getting a NPE if you're just parsing a small fragment with no start or end.


> > you must match a required end object to a required start object. You require one but make the other one optional
> 
> I am matching start and end object which are being added in our toXContent method. 

When you create the JSON you can match them. If you are reading the JSON created elsewhere you do not control that.  Someone else could be creating that JSON outside of core.

The code around the loop looks better because you're matching the start (lines 603-604) to the end (the stopping condition for the loop in 605).

But there is still an optional `START_OBJECT` present in lines 590-592 without a corresponding end object.  I still think this is wrong.

It should be legal for anyone parsing this JSON to parse their own start/end outside of your parsing.  SO taking your JSON above:
```json
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```
it should be legal to pass this (valid JSON) string as a fragment which completely encapsulates all the information in the object, lacking only its "field name" which is only relevant for `DiscoveryNodes`.
```json
  {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
This fails, however, because your fragment behavior is not just the value, it's a key-value pair expected by `DiscoveryNodes` parsing and you require the format to be:
```json
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
Essentially, the optional start object IS read when parsing a standalone object but is NOT read when parsing the specific key/value format you include in `DiscoveryNodes` which omits the end object.

So ultimately my chief complaint here is that having the "field name" as part of this object is wrong; I should be able to have a standalone object that works without needing `DiscoveryNodes` wrapped around it.

Reply:
To be more clear/succinct, your representation of `DiscoveryNode` in JSON is this (optionally wrapped in curly braces) key value pair.
```json
{
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
}
```
or this because the object wrapper is optional, but requires the method caller to know that it's expecting a field name and object:
```json
  "node_1" : {
    "name" : "name_1",
    "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
    "transport_address" : "0.0.0.0:2",
    "attributes" : {
        "custom" : "PKU"
    }
  }
```
This is not the expectation I've seen anywhere for XContentFragments which are usually standalone values (which may or may not be objects).

I believe it should either be:
{
  "name" : "name_1",
  "id" : "node_1",
  "ephemeral_id" : "3Q3xRwYKScWqBgVCrWmNCQ",
  "transport_address" : "0.0.0.0:2",
  "attributes" : {
      "custom" : "PKU"
  }
}
```
or your original format with the object wrappers mandatory (even if you keep the fragment interface).

You're mixing-and-matching pairing of start object which relies heavily on implementation details of other classes.

Reply:
@dbwiddis Thanks for being patient to write the detailed comment. Got your point on that we should be able to pass the inner nested object without the fieldname. But we need to create a parser which takes id as a parameter.
```
DiscoveryNode fromXContent(XContentParser parser, String nodeId)
```
I am assuming this will help address your concern.

Reply:
@dbwiddis Thanks for reviewing, I have added the the javadoc for this method.